br/pkg/checkpoint/manager.go

```diff
@@ -0,0 +1,414 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checkpoint
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pingcap/errors"
+	backuppb "github.com/pingcap/kvproto/pkg/brpb"
+	"github.com/pingcap/tidb/br/pkg/glue"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tidb/pkg/domain"
+	"github.com/pingcap/tidb/pkg/parser/ast"
+)
+
+type SnapshotMetaManagerT = MetaManager[
+	RestoreKeyType, RestoreValueType, RestoreValueType, CheckpointMetadataForSnapshotRestore,
+]
+type LogMetaManagerT = MetaManager[
+	LogRestoreKeyType, LogRestoreValueType, LogRestoreValueMarshaled, CheckpointMetadataForLogRestore,
+]
+
+type tickDurationConfig struct {
+	tickDurationForFlush    time.Duration
+	tickDurationForChecksum time.Duration
+	retryDuration           time.Duration
+}
+
+func DefaultTickDurationConfig() tickDurationConfig {
+	return tickDurationConfig{
+		tickDurationForFlush:    defaultTickDurationForFlush,
+		tickDurationForChecksum: defaultTickDurationForChecksum,
+		retryDuration:           defaultRetryDuration,
+	}
+}
+
+type MetaManager[K KeyType, SV, LV ValueType, M any] interface {
+	RootPath() string
```

Comment:
Looks like this is only used for formatting, how about embedding a `fmt.Stringer`?